README.md

```diff
@@ -464,6 +464,81 @@ repository, and `ZULIP_GIT_REF` to be any ref name in that repository
 
 Then stop and restart the container as described in the previous section.
 
+### Upgrading zulip/zulip-postgresql to 14
+
+These instructions assume that you have not changed the default
+Postgres data path (`/opt/docker/zulip/postgresql/data`) in your
+`docker-compose.yml`.  If you have changed it, please replace all
+occurences of `/opt/docker/zulip/postgresql/data` with your path.
+
+1. Make a backup of your Zulip Postgres data dir.
+
+2. Stop all Zulip containers, except the postgres one (e.g. use
+   `docker stop` and not `docker-compose stop`).
+
+3. Create a new (upgraded) Postgres container using a different data directory:
+
+```
+docker run -d \
+      --name postgresnew \
+      -e POSTGRES_DB=zulip \
+      -e POSTGRES_USER=zulip \
+      -e POSTGRES_PASSWORD=zulip \
+      -v /opt/docker/zulip/postgresql/new:/var/lib/postgresql/data:rw \
+      zulip/zulip-postgresql:14
+```
+
+4. Use `pg_dumpall` to dump all data from the existing Postgres container to the new Postgres container:
+
+```
+docker-compose exec database pg_dumpall -U zulip | \
+    docker exec -i postgresnew psql -U zulip
+```
+
+5. Stop and remove both Postgres containers:
+
+```
+docker-compose rm --stop database
+docker stop postgresnew
+docker rm postgresnew
+```
+
+6. Edit your `docker-compose.yml` to use the `zulip/zulip-postgresql:14` image for the `database` container.
+
+7. Replace the old Postgres data directory with upgraded data directory:
+
+```
+sudo mv /opt/docker/zulip/postgresql/data /opt/docker/zulip/postgresql/old
+sudo mv /opt/docker/zulip/postgresql/new /opt/docker/zulip/postgresql/data
+```
+
+8. Delete the old existing containers:
+
+```
+docker-compose rm
+```
+
+9. Reset the Postgres password defaults to MD5:
+
+In ``/opt/docker/zulip/postgresql/data/pg_hba.conf`` the last line should be:
+```
+host all all all md5
+```
+
+In ``/opt/docker/zulip/postgresql/data/postgresql.conf``  explicitly set:
+```
+password_encryption = md5
+```
```

Comment:
Why is this step required? I don't see a step doing the opposite, so am a little confused; is this a change we should be making in the postgres-14 container image?

That's `Dockerfile-postgresql` in zulip/zulip, BTW, in case we neglected to document that somewhere reasonable.

Reply:
As noted in the PR above:

```quote
There are two ways to resolve item 9: The easy way is as-listed, which is to just push Postgres to use MD5 auth. The less-easy way is to upgrade to SCRAM within the DB.
```

If it's a fresh install, then keeping the default (SCRAM) is perfect. If it's an upgrade, then either we need to tweak the config to keep MD5, or we have to add a few steps to make pre-upgrade changes to support SCRAM being the default.

Reply:
It seems mildly cleaner to switch to SCRAM-SHA-256.  Which we can do by re-setting the password after we've ingested all of the pg_dumpall, via:

```
# ...set $database_password from the YAML
echo "ALTER USER zulip WITH PASSWORD '$database_password';" |
    docker exec -i postgresnew psql -U zulip
```

See the top-level comment for one way to do this.  However, we're not getting a lot from the password hashing here, so I'm not too concerned with this from a security point of view -- it's just useful in keeping the postgres image as straightforward as possible.